### ISSUE TITLE
inRepoAddons from dummy app should be excluded

### ISSUE THREAD

**Issue by mehulkar** [ember-cli/ember-cli#9016]
When Ember CLI loads a project, it recursively goes through each addon's dependencies and adds them into the `PackageInfoCache`, including the in-repo addons of dependencies. It does this by looking at the `path` key in the package.json of every addon: 

https://github.com/ember-cli/ember-cli/blob/v3.15.1/lib/models/package-info-cache/index.js#L551-L559

The `path` property in the package.json of an Ember addon can point to an in-repo addon in its dummy app (because there's no package.json in the dummy app itself), so Ember CLI adds this to its PackageInfoCache, but sets a `valid` flag to false, because that directory doesn't exist:

- sets an error here: https://github.com/ember-cli/ember-cli/blob/v3.15.1/lib/models/package-info-cache/index.js#L465
- and the `valid` flag here: https://github.com/ember-cli/ember-cli/blob/v3.15.1/lib/models/package-info-cache/index.js#L509

After its done recursively discovering addons, Ember CLI then dumps all the invalid addons (i.e. the ones with the `valid: false` flag and logs a nice message here: https://github.com/ember-cli/ember-cli/blob/v3.15.1/lib/models/package-info-cache/package-info.js#L338-L369

I have a few questions:

1. Why can an addon even have an in-repo addon? Does that make any sense outside of the dummy app? Is any addon creating in-repo addons that are intended to ship with them? 
1. Is creating in-repo addons in the *dummy* app supported behavior? If not, where would it make sense to document this?
1. If both in-repo addons in addons *and* in-repo addons in the dummy app are supported, Ember CLI should probably skip the latter set instead of logging a message that says that the addon's package.json is broken.

This whole thing showed up for me because this log message was added in the 3.15 release and it was throwing an error (fixed in #8907 and will probably make it into 3.1.5.2 via #9014). Once the bug is fixed, I'll still have a log message calling this out for me, and I am not sure what to do about the warning.

---

Output from `ember version --verbose && npm --version && yarn --version`:
```
ember-cli: 3.15.1
http_parser: 2.8.0
node: 10.18.0
v8: 6.8.275.32-node.54
uv: 1.28.0
zlib: 1.2.11
brotli: 1.0.7
ares: 1.15.0
modules: 64
nghttp2: 1.39.2
napi: 5
openssl: 1.1.1d
icu: 64.2
unicode: 12.1
cldr: 35.1
tz: 2019a
os: darwin x64
6.13.4
1.21.1
```

**Comment by rwjblue** [comment]
> I have a few questions:
> 
> 1. Why can an addon even have an in-repo addon? Does that make any sense outside of the dummy app? Is any addon creating in-repo addons that are intended to ship with them?

Why wouldn't it be able to have in-repo addons? The only actual difference between stand alone addons and in-repo ones (other than the blueprint used to generate them) is how they are discovered, I would be very opposed to making them _more_ special.

> Is any addon creating in-repo addons that are intended to ship with them?

Not sure if there are any public ones, but I know of a few internal addons that do this intentionally. Most of the time, the goal is the same reason as you would use in-repo addons in an app: to have a small amount of separation but still bundle support. 

> 2. Is creating in-repo addons in the _dummy_ app supported behavior? If not, where would it make sense to document this?

Sure, why wouldn't it be?

> 3. If both in-repo addons in addons *and in-repo addons in the dummy app are supported, Ember CLI should probably skip the latter set instead of logging a message that says that the addon's package.json is broken.

Now we are getting somewhere 😸. I definitely agree that this is a bit non-obvious, and that we should make the experience better. The question is how. Here are a few options that I can think of quickly:

1. Change the `ember g in-repo` behavior to use either `link:` or `file:` paths (`link:` for `yarn` users, and `file:` for `npm` users) in `devDependencies`, and slowly begin the process of deprecating `ember-addon.paths` in `package.json` all together. (note: I prefer this option)
1. Add support for `ember-addon.developmentPaths` (basically allowing the addon to declare that some of it's inrepo paths are for development only).
1. Attempt to automatically remove any `ember-addon.paths` entries that are not included in the published tarball in a `prepublish` script (restoring in a `postpublish` script).

**Comment by mehulkar** [comment]
> > Is any addon creating in-repo addons that are intended to ship with them?
> Not sure if there are any public ones, but I know of a few internal addons that do this intentionally. 

Oh interesting. How do they include them in the published tarball? It doesn't happen by default with `npm publish` afaik.

>  I definitely agree that this is a bit non-obvious

TLDR from above: an Addon can define an IRA for *both* dummy apps and/or for shipping, but it's not super clear how that works, and it's not clear if the same addon can be used for both.

> The question is how.

Personally, I want in-repo addons to die 😛. But that aside:

- 👍 for killing `ember-addon.paths`.
- (1) from your list seems like the right direction, but the docs first need to clarify (in docs) how `dependencies` and `devDependencies` are used by Ember CLI in the first place. IIUC, ember-addons in devDependencies don't behave differently from ember-addons in dependencies, so it would be confusing to make that distinction meaningful for in-repo addons.

**Comment by rwjblue** [comment]
> How do they include them in the published tarball? It doesn't happen by default with npm publish afaik.

They put them in `lib/`, which (as long as you haven't modified `.npmignore`) will be published just like any other file.

> TLDR from above: an Addon can define an IRA for both dummy apps and/or for shipping, but it's not super clear how that works, and it's not clear if the same addon can be used for both.

Sorry, I can't tell if you are asking a question here (RE: "not super clear how that works" or "not clear if the same addon can be used for both"). Can you clarify?

> but the docs first need to clarify (in docs) how dependencies and devDependencies are used by Ember CLI in the first place

Maybe? In general, they are used **exactly** like what the `package.json` indicates. Things that are `devDependencies` are used when locally developing on the addon, things that are `dependencies` are used when consumed.

To be specific:

* in an app, there is effectively no difference between `dependencies` and `devDependencies`. Including any addon (defined by having hte `ember-addon` keyword) in either location will result in the `project.addons` including an instance of that addon.
* in addons anything with the `ember-addon` keyword that is found in dependencies of the top level will be included in the addon's own instantiated addons array
* in addons *while locally developing*, anything with the `ember-addon` keyword that is found in `devDependencies` will be included in the instantiated addons array on the `project` instance (this means the dummy app).

**Comment by locks** [comment]
@mehulkar has the conversation been resolved?

**Comment by mehulkar** [comment]
Not really,but v2 addon format with monorepo makes the problem obsolete. If we’re going to close this I’d call it a “Not to be fixed” rather than “not an issue” :)

**Comment by locks** [comment]
Thanks for the follow up! I did just that 💪
